ingest/ledgerbackend/cloud_storage_backend.go

```
@@ -0,0 +1,210 @@
+// The cloud storage service contains files generated from the ledgerExporter.
+type CloudStorageBackend struct {
+	lcmDataStore datastore.DataStore
+	storageURL   string
+}
+
+// Return a new CloudStorageBackend instance.
+func NewCloudStorageBackend(ctx context.Context, storageURL string) (*CloudStorageBackend, error) {
+	lcmDataStore, err := datastore.NewDataStore(ctx, storageURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CloudStorageBackend{lcmDataStore: lcmDataStore, storageURL: storageURL}, nil
+}
+
+// GetLatestLedgerSequence returns the most recent ledger sequence number in the cloud storage bucket.
+func (csb *CloudStorageBackend) GetLatestLedgerSequence(ctx context.Context) (uint32, error) {
+	// Get the latest parition directory from the bucket
+	directories, err := csb.lcmDataStore.ListDirectoryNames(ctx)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed getting list of directory names")
+	}
+
+	latestDirectory, err := getLatestDirectory(directories)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed getting latest directory")
+	}
+
+	// Search through the latest partition to find the latest file which would be the latestLedgerSequence
+	fileNames, err := csb.lcmDataStore.ListFileNames(ctx, latestDirectory)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed getting filenames in dir %s", latestDirectory)
+	}
+
+	latestLedgerSequence, err := getLatestFileNameLedgerSequence(fileNames, latestDirectory)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed converting filename to ledger sequence")
+	}
+
+	return latestLedgerSequence, nil
+}
+
+// GetLedger returns the LedgerCloseMeta for the specified ledger sequence number
+func (csb *CloudStorageBackend) GetLedger(ctx context.Context, sequence uint32) (xdr.LedgerCloseMeta, error) {
+	var ledgerCloseMetaBatch xdr.LedgerCloseMetaBatch
+
+	objectKey, err := GetObjectKeyFromSequenceNumber(sequence, ledgersPerFile, filesPerPartition)
```

is it worthwhile to pull up ledgersPerFile and filesPerPartition into config params on `CloudStorageBackend` struct rather than hardcode even at this early stage? Intent being to let the caller pass into NewCloudStorageBackend(), as caller would seem likely to have knowledge of how the GCS store they intend to use has been loaded in relation to these two dimensions? Has there ever been discussion on thoughts of ledger exporter emitting a separate 'config.json' file onto datastores during export..then these corresponding datastore backend implementations could retrieve it via datastore.GetFile("config.json") and dynamically obtain the same config values.

---

Yeah passing the ledgersPerFile, filesPerPartition, and probably the fileSuffix through `NewCloudStorageBackend()` makes sense.

I don't think we ever talked about an emitted `config.json` but I like that idea

---

cc: @urvisavla for the config.json idea

---

oh, cool, that suggests making the source-of-truth(SoT) for datastore schema settings(ledgers-per-file, files-per-partition, etc) be on the datastore itself, which is better than my first assumption of ledgerexporter just pushing it to datastore on every run. Also, I think `schema.json` may be more accurate name of file for this purpose, as it's not defining how to connect or anything else, just describes the schema of the data that is stored on it. 

Following the idea of datstore's copy of`schema.json` being the SoT, it enables consistency across distributed processing of ledgerexporters and backends, as both would defer to obtain the settings dynamically at runtime from datastore via GetFile. The initial requirement could be for user to place`schema.json` manually on datastore first and then all ledgerexporters and backends just expect that file to exist and read from it to obtain the schema settings and could fail fast if not found. Or we could make it more robust with some fallbacks as mentioned.

---

Discussed in team meeting and ticket created [HUBBLE-397](https://stellarorg.atlassian.net/browse/HUBBLE-397)